examples/matrix_multiplication.masm

```diff
@@ -0,0 +1,704 @@
+            ## dup.1 push.2 lte assert   # r <= rows(A)
+            ## dup.2 eq.2 assert         # columns(B)
+            ## dup.3 push.300 gt assert  # writePtr > address(C)
+            ## dup.3 push.302 lte assert # && writePtr < address(C) + total_words_in(A)
+            ## dup.4 eq.1 assert         # round_up(columns(A) / 4)
+            ## dup.5 push.100 gt assert  # readA > address(A)
+            ## dup.5 push.102 lte assert # && readA < address(A) + total_words_in(A)
+            ## dup.6 push.200 gt assert  # addrB > address(B)
+            ## dup.6 push.202 lte assert # && addrB < address(B) + total_words_in(B)
+            ## dup.7 eq.0 assert
+            ## /DEBUG
+
+            # [[i, readA, addrB, acc], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+            swapw dupw
+
+            # setup curr_col to the right column
+            # curr_col = addrB + (words * c)
+            # [[i, readA, curr_col, acc], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+            dup dup.9 mul movup.3 add movdn.2
+
+            # MATH: [[i, readA, curr_col, acc], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+            # while(i!=0)
+            dup neq.0 while.true
+
+                # [[acc, i, readA, curr_col], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                movup.3
+
+                # [[A], [acc, i, readA, curr_col], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                padw dup.6 mem_loadw
+
+                # [[B], [A], [acc, i, readA, curr_col], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                padw dup.11 mem_loadw
+
+                # note: it is okay to use this procedure for matrixes that
+                # don't have enough elements to fill the last word, because the
+                # word is padded with zeros
+                # [[acc, i, readA, curr_col], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                exec.multiply_add_word add
+
+                # update indices
+                # [[readA, curr_col+1, acc, i], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                movup.3 add.1 movup.3
+
+                # [[i, readA, curr_col, acc], [words, readA, addrB, 0], [c, r, colC, writePtr], ...]
+                movup.3
+
+                # while(i!=0)
+                sub.1 dup neq.0
+            end
+
+            # discard the auxiliary loop data, and save the computed element
+            # [[words, readA, addrB, 0], [c, r, colC, writePtr], acc, ...]
+            drop drop drop movdn.8
+
+            # count the element
+            # [[c, r, colC, writePtr], [words, readA, addrB, 0], acc, ...]
+            swapw add.1 
+
+            # while(c!=colC)
+            dup dup.3 neq
+        end
+
+        # add the padding
+        # [[c, r, colC, writePtr], [words, readA, addrB, 0], acc, ...]
+        dup push.3 u32checked_and neq.0 while.true
```

Comment:
This is the only reason why `u32` is required. I didn't find a better way of figuring out how many padding `0`s should be added to complete the row